core-bundle/src/Resources/contao/library/Contao/StringUtil.php

```diff
@@ -558,117 +559,55 @@ public static function parseSimpleTokens($strString, $arrData)
 	{
 		$strReturn = '';
 
-		$replaceTokens = static function ($strSubject) use ($arrData)
+		// Make sure we use proper token names so the expression language can handle them
+		$arrTokenKeyHashes = array();
+
+		foreach ($arrData as $k => $v)
+		{
+			$hashedKey = 't' . sha1($k); // Make sure we prefix with a letter, otherwise it's not considered a valid variable
+
+			$arrTokenKeyHashes[$k] = $hashedKey;
+			$arrData[$hashedKey] = $v;
+			unset($arrData[$k]);
+		}
+
+		$replaceTokens = static function ($strSubject) use ($arrTokenKeyHashes, $arrData)
 		{
 			// Replace tokens
 			return preg_replace_callback
 			(
 				'/##([^=!<>\s]+?)##/',
-				static function (array $matches) use ($arrData)
+				static function (array $matches) use ($arrTokenKeyHashes, $arrData)
 				{
-					if (!\array_key_exists($matches[1], $arrData))
+					if (!\array_key_exists($matches[1], $arrTokenKeyHashes))
 					{
 						System::getContainer()
 							->get('monolog.logger.contao')
-							->log(LogLevel::INFO, sprintf('Tried to parse unknown simple token "%s".', $matches[1]))
-						;
+							->log(LogLevel::INFO, sprintf('Tried to parse unknown simple token "%s".', $matches[1]));
 
 						return '##' . $matches[1] . '##';
 					}
 
-					return $arrData[$matches[1]];
+					return $arrData[$arrTokenKeyHashes[$matches[1]]];
 				},
 				$strSubject
 			);
 		};
 
-		$evaluateExpression = static function ($strExpression) use ($arrData)
-		{
-			if (!preg_match('/^([^=!<>\s]+) *([=!<>]+)(.+)$/s', $strExpression, $arrMatches))
-			{
-				return false;
-			}
-
-			$strToken = $arrMatches[1];
-			$strOperator = $arrMatches[2];
-			$strValue = trim($arrMatches[3], ' ');
-
-			if (!\array_key_exists($strToken, $arrData))
-			{
-				System::getContainer()
-					->get('monolog.logger.contao')
-					->log(LogLevel::INFO, sprintf('Tried to evaluate unknown simple token "%s".', $strToken))
-				;
+		$el = new ExpressionLanguage();
 
-				return false;
-			}
-
-			$varTokenValue = $arrData[$strToken];
+		$evaluateExpression = static function ($strExpression) use ($el, $arrTokenKeyHashes, $arrData)
+		{
+			// Replace with hashes
+			$strExpression = str_replace(array_keys($arrTokenKeyHashes), array_values($arrTokenKeyHashes), $strExpression);
```

Comment:
I don’t think this can work reliably. E.g. `{if firstname == "myname"}` with the data `['name' => 'myname mylastname', 'firstname' => 'myname']` breaks.

Reply:
I see, I'm not sure how to fix this, though. It's only the special chars in the variables/tokens which are not supported by the EL (e.g. the `#`). Not sure why one would ever use those...